packages/datadog-instrumentations/src/playwright.js

```diff
@@ -0,0 +1,236 @@
+
+  if (_type === 'beforeAll' || _type === 'afterAll') {
+    return
+  }
+
+  const testResult = results[results.length - 1]
+  const testAsyncResource = testToAr.get(test)
+  testAsyncResource.runInAsyncScope(() => {
+    testFinishCh.publish({ testStatus, steps: testResult.steps, error })
+  })
+
+  if (!testSuiteToTestStatuses.has(testSuiteAbsolutePath)) {
+    testSuiteToTestStatuses.set(testSuiteAbsolutePath, [testStatus])
+  } else {
+    testSuiteToTestStatuses.get(testSuiteAbsolutePath).push(testStatus)
+  }
+
+  remainingTestsByFile[testSuiteAbsolutePath] = remainingTestsByFile[testSuiteAbsolutePath]
+    .filter(currentTest => currentTest !== test)
+
+  if (!remainingTestsByFile[testSuiteAbsolutePath].length) {
+    const testStatuses = testSuiteToTestStatuses.get(testSuiteAbsolutePath)
+
+    let testSuiteStatus = 'pass'
+    if (testStatuses.some(status => status === 'fail')) {
+      testSuiteStatus = 'fail'
+    } else if (testStatuses.every(status => status === 'skip')) {
+      testSuiteStatus = 'skip'
+    }
+
+    const testSuiteAsyncResource = testSuiteToAr.get(testSuiteAbsolutePath)
+    testSuiteAsyncResource.runInAsyncScope(() => {
+      testSuiteFinishCh.publish(testSuiteStatus)
+    })
+  }
+}
+
+function dispatcherRunWrapper (run) {
+  return function () {
+    remainingTestsByFile = getTestsBySuiteFromTestsById(this._testById)
+    return run.apply(this, arguments)
+  }
+}
+
+function dispatcherHook (dispatcherExport) {
+  shimmer.wrap(dispatcherExport.Dispatcher.prototype, 'run', dispatcherRunWrapper)
+  shimmer.wrap(dispatcherExport.Dispatcher.prototype, '_createWorker', createWorker => function () {
+    const dispatcher = this
+    const worker = createWorker.apply(this, arguments)
+
+    worker.process.on('message', ({ method, params }) => {
+      if (method === 'testBegin') {
+        const { test } = dispatcher._testById.get(params.testId)
+        testBeginHandler(test)
+      } else if (method === 'testEnd') {
+        const { test } = dispatcher._testById.get(params.testId)
+
+        const { results } = test
+        const testResult = results[results.length - 1]
+
+        testEndHandler(test, STATUS_TO_TEST_STATUS[testResult.status], testResult.error)
+      }
+    })
+
+    return worker
+  })
+  return dispatcherExport
+}
+
+function dispatcherHookNew (dispatcherExport) {
+  shimmer.wrap(dispatcherExport.Dispatcher.prototype, 'run', dispatcherRunWrapper)
+  shimmer.wrap(dispatcherExport.Dispatcher.prototype, '_createWorker', createWorker => function () {
+    const dispatcher = this
+    const worker = createWorker.apply(this, arguments)
+
+    worker.on('testBegin', ({ testId }) => {
+      const { test } = dispatcher._testById.get(testId)
+      testBeginHandler(test)
+    })
+    worker.on('testEnd', ({ testId, status, errors }) => {
+      const { test } = dispatcher._testById.get(testId)
+
+      testEndHandler(test, STATUS_TO_TEST_STATUS[status], errors[0])
```

Comment:
Has `errors` a value when the test is success? Should we protect it to be sure that it works if they change the implementation in the future?
```suggestion
      testEndHandler(test, STATUS_TO_TEST_STATUS[status], errors && errors[0])
```

Reply:
yes, it should always be defined, at least for the versions we support right now. It's a good idea to check though, so I'll do it in a follow up PR

Reply:
I'm working on https://github.com/DataDog/dd-trace-js/pull/2725 right now, which will include some playwright changes, so I'll add this suggestion there